src/snowflake/snowpark/_internal/analyzer/metadata_utils.py

```diff
@@ -0,0 +1,61 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+from typing import List, Optional, Tuple
+
+from snowflake.snowpark._internal.analyzer.expression import Attribute
+from snowflake.snowpark._internal.analyzer.snowflake_plan_node import Limit, LogicalPlan
+
+
+def infer_metadata(
+    source_plan: LogicalPlan,
+) -> Tuple[Optional[List[Attribute]], Optional[List[str]]]:
+    """
+    Infer metadata from the source plan.
+    Returns the metadata including attributes (schema) and quoted identifiers (column names).
+    """
+    from snowflake.snowpark._internal.analyzer.select_statement import (
+        Selectable,
+        SelectStatement,
+    )
+    from snowflake.snowpark._internal.analyzer.snowflake_plan import SnowflakePlan
+    from snowflake.snowpark._internal.analyzer.unary_plan_node import (
+        Filter,
+        Sample,
+        Sort,
+    )
+
+    attributes = None
+    quoted_identifiers = None
+    # If source_plan is a LogicalPlan, SQL simplfiier is not enabled
+    # so we can try to infer the metadata from its child (SnowflakePlan)
+    # When source_plan is Filter, Sort, Limit, Sample, metadata won't be changed
+    # so we can use the metadata from its child directly
+    if isinstance(source_plan, (Filter, Sort, Limit, Sample)):
+        if isinstance(source_plan.child, SnowflakePlan):
+            attributes = source_plan.child._attributes
+            quoted_identifiers = source_plan.child._quoted_identifiers
+    # If source_plan is a SelectStatement, SQL simplifier is enabled
+    elif isinstance(source_plan, SelectStatement):
+        # When source_plan._snowflake_plan is not None, `get_snowflake_plan` is called
+        # to create a new SnowflakePlan and `infer_metadata` is already called on the new plan.
+        if source_plan._snowflake_plan is not None:
+            attributes = source_plan._snowflake_plan._attributes
+            quoted_identifiers = source_plan._snowflake_plan._quoted_identifiers
```

Comment:
it seems the quoted identifiers currently is not used anywhere, and there is no actual inferring yet, by inferring i mean extracting the name from the projection. maybe we can do it along with when we actual use it in a different pr

Reply:
sure, let's add it in another PR